### PR TITLE
DHT interfaces 

### PR DESCRIPTION
**Pull request by coyle**
The interface definitions for our DHT

### CODE REVIEW

File: pkg/kademlia/kademlia.go [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+	"context"
+
+	"storj.io/storj/protos/overlay"
+)
+
+type nodeID string
+
+// DHT is the interface for the DHT in the Storj network
+type DHT interface {
+	GetNodes(ctx context.Context) []overlay.Node
```
Comment (jtolds) [inline pkg/kademlia/kademlia.go]: this returns, what, all nodes? all 100k? everything in the routing table? the routing table might be big

File: pkg/kademlia/kademlia.go [hunk 2]
```diff
@@ -0,0 +1,36 @@
+package kademlia
+
+import (
+	"context"
+
+	"storj.io/storj/protos/overlay"
+)
+
+type nodeID string
```
Comment (jtolds) [inline pkg/kademlia/kademlia.go]: should probably be exported?

File: pkg/kademlia/kademlia.go [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+// DHT is the interface for the DHT in the Storj network
+type DHT interface {
+	GetNodes(ctx context.Context) []overlay.Node
+	GetRoutingTable(ctx context.Context) RoutingTable
+	Bootstrap(ctx context.Context) error
```
Comment (jtolds) [inline pkg/kademlia/kademlia.go]: good job on the ctx!

File: pkg/kademlia/kademlia.go [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+	Ping(ctx context.Context, node overlay.Node) overlay.Node
```
Comment (jtolds) [inline pkg/kademlia/kademlia.go]: we probably don't need Ping exposed

File: pkg/kademlia/kademlia.go [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+	FindNode(ctx context.Context, ID nodeID) overlay.Node
+	FindValue(ctx context.Context)
```
Comment (jtolds) [inline pkg/kademlia/kademlia.go]: probably should take an ID (a value id), and return a value

File: pkg/kademlia/kademlia.go [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+}
+
+// RoutingTable contains information on nodes we have locally
+type RoutingTable interface {
+	GetBuckets() []*Bucket
+	GetBucket(id string) (ok bool, bucket Bucket)
```
Comment (jtolds) [inline pkg/kademlia/kademlia.go]: GetBuckets() could return lots of things. maybe make it take a callback per bucket?
returning ok bool first before bucket seems unusual

overall i'd probably make this a bit more high level - 

```
type RoutingTable interface {
  // local params
  LocalId() nodeId
  K() int
  CacheSize() int

  GetBucket(id string) (bucket Bucket, ok bool)
  GetBuckets() []Bucket

  FindNear(id nodeId, limit int) ([]overlay.Node, error)
  
  ConnectionSuccess(id string, address overlay.NodeAddress)
  ConnectionFailed(id string, address overlay.NodeAddress)

  // these are for refreshing
  SetBucketTimestamp(id string, now time.Time) error
  GetBucketTimestamp(id string, bucket Bucket) (time.Time, error)
}
```

File: pkg/kademlia/kademlia.go [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+
+// Bucket is a set of methods to act on kademlia k buckets
+type Bucket interface {
+	MoveToTail(ID nodeID) error
+	MoveToHead(ID nodeID) error
+	Has(ID nodeID) bool
+	Add(ID nodeID) error
+	Remove(ID nodeID) error
+	Get(ID nodeID) overlay.Node
+	Len() int
```
Comment (jtolds) [inline pkg/kademlia/kademlia.go]: with the methods for the routing table above, the stuff you need in the Bucket are very little, maybe just 

```
type Bucket interface {
  Routing() []overlay.Node
  Cache() []overlay.Node
  Midpoint() string
}
```

Comment (coyle) [reply pkg/kademlia/kademlia.go]: I was basing a lot of those off the paper where it describes adding/removing elements from a bucket. You thinking those shouldn't be exposed methods ?

File: pkg/kademlia/kademlia.go (hunk 4 above)
Comment (jtolds) [reply pkg/kademlia/kademlia.go]: :+1:

File: pkg/kademlia/kademlia.go (hunk 7 above)
Comment (jtolds) [reply pkg/kademlia/kademlia.go]: well, the only time elements are added or removed are from these two and only these two callback methods:

```
  ConnectionSuccess(id string, address overlay.NodeAddress)
  ConnectionFailed(id string, address overlay.NodeAddress)
```

so i think it's probably less error prone to have the ability to manipulate the bucket in just those two places? i dunno. 

maybe we need a high level interface and a low level interface, and the high level is implemented in terms of the low level?

File: pkg/kademlia/kademlia.go (hunk 1 above)
Comment (dylanlott) [reply pkg/kademlia/kademlia.go]: Pagination makes the most sense to me.

File: pkg/kademlia/kademlia.go (hunk 6 above)
Comment (coyle) [reply pkg/kademlia/kademlia.go]: I'm not understanding 
```
GetBuckets(cb func(Bucket) (keep_going bool))
```

Why would get buckets just return an array of buckets ?

File: pkg/kademlia/kademlia.go (hunk 7 above)
Comment (coyle) [reply pkg/kademlia/kademlia.go]: OK I think I grasp what you are saying. 

The logic of dropping an address, adding an address, splitting the buckets etc. will all be implemented directly within these two methods.

File: pkg/kademlia/kademlia.go (hunk 6 above)
Comment (jtolds) [reply pkg/kademlia/kademlia.go]: sorry, you're totally right, there's a fixed number of DHT buckets. it's not actually capped at 256 (there's edge cases where a non-rightmost bucket might split in the paper), but it's pretty close, and that totally doesn't need pagination

**Pull request merged**